src/BinSkim.Rules/PERules/BA2006.BuildWithSecureTools.cs

```diff
@@ -66,6 +66,29 @@ public IEnumerable<IOption> GetOptions()
             new PerLanguageOption<AdvancedMitigations>(
                 AnalyzerName, nameof(AdvancedMitigationsEnforced), defaultValue: () => AdvancedMitigations.None);
 
+        public override void Initialize(BinaryAnalyzerContext context)
+        {
+            if (context.Policy == null) { return; }
+
+            StringToVersionMap currentMinimumToolVersions = context.Policy.GetProperty(MinimumToolVersions, cacheDefault: false);
+            StringToVersionMap defaultMinimumToolVersions = BuildMinimumToolVersionsMap();
+
+            // If user only overwrite part of the settings, currentMinimumToolVersions will only have those settings.
+            // This is to make sure other default settings are applied as well.
+            foreach (KeyValuePair<string, Version> defaultMinimumToolVersion in defaultMinimumToolVersions)
+            {
+                currentMinimumToolVersions.TryAdd(defaultMinimumToolVersion.Key, defaultMinimumToolVersion.Value);
```

Comment:
>currentMinimumToolVersions.TryAdd(defaultMinimumToolVersion.Key, defaultMinimumToolVersion.Value); [](http://example.com/codeflow?start=16&length=98)

@michaelcfanning , If we do this, the user would not be able to remove and use the min version since we are always defaulting to our current mapping.

Probably here we would have to set to new Version()

Reply:
the cases i thought:
1. user is using default -> policy should be the same as the default mapping
2. user customized one version -> that version should be customized
3. user removed one version -> that version should show as 0.0.0.0

Reply:
Good test cases but your expected outcomes aren't quite right. In #3, if a user removes the version, that value should return to the default, which is the value that's encoded in the rule. This is the behavior of the configuration: if you explicitly refer to it, you can override it. If you do *not* refer to it, you get the built-in default behavior.

This is why Shaopeng's fix is correct. At runtime, he traverses the configuration that comes in and re-adds any properties that are absent.

What customers needs to do if they want to, say, permit all C compiler versions, is to leave an explicit reference to the compiler and to explicitly set its value to zero. After this, all compiler references to that version will be accepted.